// @azure/storage-queue/dist-esm/storage-queue/src/index.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { RestError } from "@azure/core-rest-pipeline";
export * from "./AccountSASPermissions";
export * from "./AccountSASResourceTypes";
export * from "./AccountSASServices";
export { generateAccountSASQueryParameters, } from "./AccountSASSignatureValues";
export * from "../../storage-blob/src/credentials/AnonymousCredential";
export * from "../../storage-blob/src/credentials/Credential";
export * from "../../storage-blob/src/credentials/StorageSharedKeyCredential";
export { BaseRequestPolicy } from "../../storage-blob/src/policies/RequestPolicy";
export { Pipeline, isPipelineLike, newPipeline, StorageOAuthScopes, } from "./Pipeline";
export * from "../../storage-blob/src/policies/AnonymousCredentialPolicy";
export * from "../../storage-blob/src/policies/CredentialPolicy";
export * from "../../storage-blob/src/StorageRetryPolicyFactory";
export * from "../../storage-blob/src/policies/StorageSharedKeyCredentialPolicy";
export * from "../../storage-blob/src/StorageBrowserPolicyFactory";
export { StorageQueueAudience, getQueueServiceAccountAudience } from "./models";
export * from "./QueueClient";
export * from "./QueueSASPermissions";
export { generateQueueSASQueryParameters, } from "./QueueSASSignatureValues";
export * from "./QueueServiceClient";
export * from "./SASQueryParameters";
export * from "./generatedModels";
export { RestError };
export { logger } from "./log";
